nixos/prosody: fix broken tests

erosennin opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Prosody tests are broken after the [slixmpp update](https://github.com/NixOS/nixpkgs/commit/47dd6e864ffa383a58936adcd82d0c7380f803f3):

https://hydra.nixos.org/build/136781589
https://hydra.nixos.org/build/136779667

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

erosennin committed:
nixos/prosody: fix broken tests

picnoir reviewed (changes requested):
Thanks for the heads up!

We want the muc join operation to be blocking. If it's not, we won't wait for the MUC test to be run. Looking at https://github.com/poezio/slixmpp/commit/c7d87a27e18aa15d317fb8b8c262f0567f3d1253 , it seems like we now want to use the `join_muc_wait` function instead of the `join_muc` + `wait=true`.

The `CthonTest` class is run asynchronously. I did not find a way to exit with the appropriate exitcode, hence the grep. By removing that grep call, the test will never fail. To the point: I assume you removed it from here because `join_muc` was not executed hence  "XMPP SCRIPT TEST SUCCESSS" never printed. This failure is however legit :)

NinjaTrappeur commented:
TL;DR: 

```diff
-        self.plugin['xep_0045'].join_muc('testMucRoom', 'cthon98')
+        self.plugin['xep_0045'].join_muc_wait('testMucRoom', 'cthon98')
```

and reverting the grep diff line 88 should do the trick.

erosennin commented:
> it seems like we now want to use the `join_muc_wait`

That was my first thought, but `join_muc_wait()` is not in the [released version](https://lab.louiz.org/poezio/slixmpp/-/blob/slix-1.7.0/slixmpp/plugins/xep_0045/muc.py) yet. Seems like there is no easy way to wait for actually joining the chat with silxmpp 1.7.0. Perhaps we should return to it later after 1.8.0 is released.

> The CthonTest class is run asynchronously. I did not find a way to exit with the appropriate exitcode, hence the grep.

Hmmm, you are right. I removed the grep mainly because it hid all logging messages and tracebacks and made the test so hard to debug. Added a workaround instead. Also made sure that the script does not hang indefinitely if an error happens.

picnoir reviewed (approved):
Nice!

Agree, I did not find any solution either regarding the `join_muc` method. Let's wait for 1.8.

Thanks for having taken the time for not only fixing the test but also greatly improving the debug log situation. kudos!
